### PR TITLE
Add an NVM Access Test for all counters in NVM

### PR DESCRIPTION
**Pull request by yashika-batra**

### CODE REVIEW

**Comment by yashika-batra** [comment]
Waiting to merge #227  : Once asyncio/test selection is merged into main, we can merge main with this branch and then merge the NVM access test specific updates.

File: state_machine/applications/system-check/main.py [hunk 1]
```diff
@@ -63,6 +66,7 @@
     ("Coil Driver Test", "coil", tests.coil_test, True),
     ("Burnwire Test", "burn", tests.burnwire_test, False),
     ("I2C_Scan", "i2c", tests.i2c_scan, False),
+    ("NVM Test", "nvm", tests.nvm_access_test, True)
```
Comment (thetazero) [inline state_machine/applications/system-check/main.py:69]: Comma at the end of this line to keep commits small

File: state_machine/applications/system-check/tests/nvm_access_test.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+nvm_counters = {"c_boot": cubesat.c_boot,
+                "c_state_err": cubesat.c_state_err,
+                "c_vbus_rst": cubesat.c_vbus_rst,
+                "c_deploy": cubesat.c_deploy,
+                "c_downlink": cubesat.c_downlink,
+                "c_logfail": cubesat.c_logfail}
+nvm_counters_items = sorted(nvm_counters.items())
+nvm_counters_values = sorted(nvm_counters.values())
+
+def verify_bits(counterstr, counter):
```
Comment (thetazero) [inline state_machine/applications/system-check/tests/nvm_access_test.py:12]: This function feels pretty hardcoded, and will break if we change the sizes of our NVM counters.
```python
    def __init__(self,num_bits,register,lowest_bit):
        self.maxval = (1 << num_bits)-1
        self.bit_mask = self.maxval << lowest_bit
        self.lowest_bit=lowest_bit
        self.byte=register
```
Looking at the source code of multiBitFlag we should be able to extract the max value automatically.

File: state_machine/applications/system-check/tests/nvm_access_test.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    else:
+        return verify_eight_bits(counter)
+
+def verify_four_bits(counter):
+    return 0 <= counter < 16
+
+def verify_eight_bits(counter):
+    return 0 <= counter < 256
+
+def incr_logfail_count(cubesat):
```
Comment (thetazero) [inline state_machine/applications/system-check/tests/nvm_access_test.py:26]: This and the following function can just be replaced by their contents:
```python
cubesat.c_logfail += 1
# and 
cubesat.c_logfail = 0
```

File: state_machine/applications/system-check/tests/nvm_access_test.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    if counter_inrange:
+        counter_inrange_string = "All existing counters are in range."
+    else:
+        counter_inrange_string = "The following counters are not in range:"
+
+    for i in range(len(nvm_counters_values)):
+        # if all counters are accessible, we'll never reach this
+        # we only add to counter_access_string if something is inaccessible
+        if not nvm_counters_exist[i]:
+            counter_access_string += nvm_counters_items[i][0] + ";"
```
Comment (thetazero) [inline state_machine/applications/system-check/tests/nvm_access_test.py:61]: probably want a space after the ";"

File: state_machine/applications/system-check/tests/nvm_access_test.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        # if the counter exists, verify its values
```
Comment (thetazero) [inline state_machine/applications/system-check/tests/nvm_access_test.py:62]: This comment is misleading, we do not verify its values at this point.

File: state_machine/applications/system-check/tests/nvm_access_test.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    """ increment logfail count in non-volatile memory (nvm) """
+    cubesat.c_logfail += 1
+
+def reset_logfail_count(cubesat):
+    """ reset logfail count in non-volatile memory (nvm) """
+    cubesat.c_logfail = 0
+
+async def run(result_dict):
+    """
+    For each i2c device, print addresses of connected devices
```
Comment (thetazero) [inline state_machine/applications/system-check/tests/nvm_access_test.py:36]: misleading doc string

**Review by thetazero — CHANGES_REQUESTED** [review]
Really like the direction this is heading in.

Two things are missing however:
- The single bit flags like f_deploy or f_burn.
- Checks to ensure the multi bit flags do not interfere with each other or reference the same region on non volatile memory. You can test this by setting all multi bit flags to their max value, then for each multibit flag change it to 0, and ensure that none of the other multi bit flags were changed.

**Comment by yashika-batra** [comment]
Thanks for the suggestions! I've added tests for flags, as well as the test you mentioned about counter memory interference. Let me know if there's anything else you have in mind!

File: state_machine/applications/system-check/tests/nvm_access_test.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            # if all counters are in range, we'll never reach this
+            # we only add to counter_flag_inrange_string if something is out of range
+            if not nvm_counters_inrange[i]:
+                counter_flag_inrange_string += nvm_counters_items[i][0] + "; "
+
+    for i in range(len(nvm_flags_values)):
+        # if all flags are accessible, we'll never reach this
+        # we only add to counter_flag_access_string if something is inaccessible
+        if not nvm_flags_exist[i]:
+            counter_flag_access_string += nvm_flags_items[i][0] + "; "
```
Comment (jbwillis) [inline state_machine/applications/system-check/tests/nvm_access_test.py:117]: This is definitely an issue - `i` does not correspond between `nvm_flags_values` and `nvm_flags_items`

File: state_machine/drivers/pycubedmini/lib/logging.py [hunk 8]
```diff
@@ -251,7 +251,7 @@ def clear_storage(folder):
     if INFO_FILENAME in listdir(f"{sd_card_directory}/{folder}"):
         remove(info_filename_dir)
 
-    rmdir(f"{sd_card_directory}/{folder}")
+    # rmdir(f"{sd_card_directory}/{folder}")
```
Comment (jbwillis) [inline state_machine/drivers/pycubedmini/lib/logging.py:254]: I don't think this should be part of this PR

File: state_machine/drivers/pycubedmini/lib/pycubed.py [hunk 9]
```diff
@@ -425,16 +437,20 @@ def timeon(self):
...
-
-    def incr_logfail_count(self):
-        """ increment logfail count in non-volatile memory (nvm) """
-        self.c_logfail += 1
+    def reset_nvm(self):
+        """ reset all flags and counters in non volatile memory """
+        # Define NVM flags
+        self.f_contact = 0
+        self.f_burn = 0
+        self.f_free1 = 0
```
Comment (jbwillis) [inline state_machine/drivers/pycubedmini/lib/pycubed.py:445]: what are `f_free1` and `f_free2`?

File: state_machine/drivers/pycubedmini/lib/pycubed.py [hunk 10]
```diff
@@ -123,6 +134,7 @@ def __init__(self):
         self.drv_y
         self.drv_z
         self.burnwire1
+        self.c_boot += 1  # increment boot count
```
Comment (jbwillis) [inline state_machine/drivers/pycubedmini/lib/pycubed.py:137]: This should happen immediately after setting `BOOTTIME` (line 113)

**Review by jbwillis — CHANGES_REQUESTED** [review]

File: state_machine/applications/system-check/main.py [hunk 11]
```diff
@@ -29,6 +34,8 @@
     ("Coil Driver Test", "coil", tests.coil_test, True),
     ("Burnwire Test", "burn", tests.burnwire_test, False),
     ("I2C_Scan", "i2c", tests.i2c_scan, False),
+    ("NVM Test", "nvm", tests.nvm_access_test, True),
+    # ("Logging Infrastructure Test", "log", tests.logging_infrastructure_test, True),
```
Comment (jbwillis) [inline state_machine/applications/system-check/main.py:38]: delete this line (logging infrastructure...)

File: state_machine/drivers/pycubedmini/lib/pycubed.py [hunk 12]
```diff
@@ -425,16 +434,18 @@ def timeon(self):
...
-    def reset_boot_count(self):
-        """ reset boot count in non-volatile memory (nvm) """
-        self.c_boot = 0
-
-    def incr_logfail_count(self):
-        """ increment logfail count in non-volatile memory (nvm) """
-        self.c_logfail += 1
+    def reset_nvm(self):
+        """ reset all flags and counters in non volatile memory """
+        # Define NVM flags
```
Comment (jbwillis) [inline state_machine/drivers/pycubedmini/lib/pycubed.py:439]: Define -> Reset

File: state_machine/drivers/pycubedmini/lib/pycubed.py [hunk 13]
```diff
@@ -425,16 +434,18 @@ def timeon(self):
...
+        self.f_contact = 0
+        self.f_burn = 0
 
-    def reset_logfail_count(self):
-        """ reset logfail count in non-volatile memory (nvm) """
+        # Define NVM counters
```
Comment (jbwillis) [inline state_machine/drivers/pycubedmini/lib/pycubed.py:443]: Define -> Reset

File: state_machine/applications/system-check/main.py [hunk 14]
```diff
@@ -76,3 +83,8 @@ async def main_test():
 
 tasko.add_task(main_test(), 1)
 tasko.run()
+
+nvm_reset = input(f"\n\nWould you like to reset non-volatile memory? Select {bold}(y){normal} for yes," +
```
Comment (jbwillis) [inline state_machine/applications/system-check/main.py:87]: I'd rather not have this here - I think it belongs in the nvm test file.

When the NVM test runs, all of the NVM registers are changed. Do they get reset at the end of the test?
Let's just have the NVM test reset all of the NVM as it completes.

**Review by jbwillis — CHANGES_REQUESTED** [review]

File: state_machine/applications/system-check/main.py [hunk 15]
```diff
@@ -29,6 +34,8 @@
     ("Coil Driver Test", "coil", tests.coil_test, True),
     ("Burnwire Test", "burn", tests.burnwire_test, False),
     ("I2C_Scan", "i2c", tests.i2c_scan, False),
+    ("NVM Test", "nvm", tests.nvm_access_test, True),
```
Comment (jbwillis) [inline state_machine/applications/system-check/main.py:37]: Per my comment below about resetting the NVM, I think this should be renamed to "Reset and test NVM"

Comment (jbwillis) [reply state_machine/applications/system-check/main.py:37]: It should be clear here that the test modifies the NVM

**Review by jbwillis — CHANGES_REQUESTED** [review]

**Pull request closed without merging**